staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root/discovery.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package root
+
+import (
+	"net/http"
+
+	restful "github.com/emicklei/go-restful/v3"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	unaggregated "k8s.io/apiserver/pkg/endpoints/discovery"
+	"k8s.io/apiserver/pkg/endpoints/discovery/aggregated"
+	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+)
+
+const APIGroupPrefix = "/apis"
+
+type RootManager interface {
+	WebService() *restful.WebService
+}
+
+// RootAPIsHandler creates a webservice serving /apis
+type RootAPIsHandler struct {
+	serializer               runtime.NegotiatedSerializer
+	discoveryGroupManager    unaggregated.GroupManager
+	discoveryResourceManager aggregated.ResourceManager
+}
+
+func NewRootAPIsHandler(serializer runtime.NegotiatedSerializer, discoveryGroupManager unaggregated.GroupManager, discoveryResourceManager aggregated.ResourceManager) *RootAPIsHandler {
+	return &RootAPIsHandler{
+		serializer:               serializer,
+		discoveryGroupManager:    discoveryGroupManager,
+		discoveryResourceManager: discoveryResourceManager,
+	}
+}
+
+func (s *RootAPIsHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AggregatedDiscoveryEndpoint) {
+		// TODO(jefftree): Error handling
+		mediaType, _ := negotiation.NegotiateMediaTypeOptions(req.Header.Get("Accept"), s.serializer.SupportedMediaTypes(), aggregated.DiscoveryEndpointRestrictions)
+		// mediaType.Convert looks at the request accept headers and is used to control whether the discovery document will be aggregated.
+		if aggregated.IsAggregatedDiscoveryGVK(mediaType.Convert) {
```

Comment:
I might suggest this as a switch statement and include the default behavior inside the block (and have a return statement) so that when we're done we delete the old loop.  makes it a bit clearer that we're walking through a set of preferred options.

It's ok to break negotiation into two parts - selecting a preferred supported option, and then dealing with the consequences of that choice (even if the impl is different).

Reply:
I was somewhat confused reading the code with the switch:
```golang
case true:
...
default:
...
```
🤔